### ISSUE TITLE
SteamNetworkingSockets support with the new flat interface in 1.48

### ISSUE THREAD

**Issue by Frooxius** [rlabrecque/Steamworks.NET#341]
Hello!

I'd like to ask if the SteamNetworkingSockets API is planned to be included as part of Steamworks.NET, with the addition of the flat C interface in the version 1.48 (https://github.com/ValveSoftware/GameNetworkingSockets/issues/70).

The topic mentions that Valve is working with you on the implementation, so I wanted to ask if that's indeed coming and if there's any rough ETA.

Thank you for your time!

**Comment by shiftyscales** [comment]
I'm also curious to know the answer to this, as it is necessary for a project I am working on, @rlabrecque. Thanks as well for your time.

**Comment by RileyGuy** [comment]
@rlabrecque I would also like to know the answer to this as I use the project that froox and shifty are part of and I'm having issues with the current network implementation that's in place at the moment.

**Comment by rlabrecque** [comment]
Some work has started on this last weekend, should be able to get something in this weekend!

**Comment by shiftyscales** [comment]
That's fantastic to hear. Thanks for the update, @rlabrecque.

**Comment by rlabrecque** [comment]
Initial support is in as of https://github.com/rlabrecque/Steamworks.NET/commit/6700235f85eb14b08bd9e4d7de90e60492bf00b0

Note that it's likely very rough, I haven't started trying to use this yet, but it's all there now!

I probably won't get a chance to actually use this for a bit, so if anyone tries it out please let me know any pain points you hit.

**Comment by Frooxius** [comment]
Thank you so much for working on this! I have given it a try, but ran into an issue, using any of the functions throws following exception:

`System.TypeLoadException: 'Could not load type 'OptionValue' from assembly 'Steamworks.NET, Version=14.0.0.0, Culture=neutral, PublicKeyToken=null' because it contains an object field at offset 0 that is incorrectly aligned or overlapped by a non-object field.'`

It seems that it doesn't like this struct:

```CSharp
		[StructLayout(LayoutKind.Explicit)]
		public struct OptionValue
		{
			[FieldOffset(0)]
			public int m_int32;

			[FieldOffset(0)]
			public long m_int64;

			[FieldOffset(0)]
			public float m_float;

			[FieldOffset(0)]
			public string m_string; // Points to your '\0'-terminated buffer

			[FieldOffset(0)]
			public IntPtr m_functionPtr;
		}
```

Particularly the string member. If I comment it out then it works fine, but it's not possible to use string based options.

Using an IntPtr to explicitly marshaled string (like with Marshal.StringToHGlobalAnsi) should work there, but adds extra work on the API user.

Is there a better solution to this?

**Comment by Frooxius** [comment]
I found another issue with the wrapper, on the function ReceiveMessagesOnPollGroup and ReceiveMessagesOnConnection.

They use `out SteamNetworkingMessage_t ppOutMessages` for the messages, but the native function seems to instead fill an array of pointers to the messages.

I've changed the native method import to this:
```CSharp
[DllImport(NativeLibraryName, EntryPoint = "SteamAPI_ISteamNetworkingSockets_ReceiveMessagesOnPollGroup", CallingConvention = CallingConvention.Cdecl)]
		public static extern int ISteamNetworkingSockets_ReceiveMessagesOnPollGroup(IntPtr instancePtr, HSteamNetPollGroup hPollGroup, IntPtr[] ppOutMessages, int nMaxMessages);```

And updated the methods appropriately and then used `Marshal.PtrToStructure<SteamNetworkingMessage_t>(ptr)` to get the structs, which worked and I managed to successfully send and receive data!

Adding a method that wraps this behavior around could help. The ValveSockets project provides methods for the open source version of the networking library, which I think could be adapted: https://github.com/nxrighthere/ValveSockets-CSharp

**Comment by rlabrecque** [comment]
Thanks for having a look dude! Sorry I didn't get to test it out that much!

> Particularly the string member. If I comment it out then it works fine, but it's not possible to use string based options.
>
> Using an IntPtr to explicitly marshaled string (like with Marshal.StringToHGlobalAnsi) should work there, but adds extra work on the API user.
>
> Is there a better solution to this?

Hmm yeah that is definitely supposed to be an IntPtr

https://github.com/rlabrecque/SteamworksSDK/blob/master/public/steam/steamnetworkingtypes.h#L1146

Maybe IntPtr with a Getter/Setter like we're looking at in https://github.com/rlabrecque/Steamworks.NET/pull/347 ?

**Comment by rlabrecque** [comment]
https://github.com/rlabrecque/Steamworks.NET/commit/455741e73e4687d71780374a763e9a5e11539672

https://github.com/rlabrecque/Steamworks.NET/commit/4aa0ee8eed6f63b86d21c32f72b46dc28a61f467

**Comment by Frooxius** [comment]
Thank you so much! And no worries, you did pretty much the brunt of it which helped a lot, I'm happy to help test it out. 

There's one more thing that I had to modify on my end. The `m_pfnRelease` member of SteamNetworkingMessage_t technically doesn't point to the message itself, but rather a function, the ValveSockets seems to be using it as a shortcut for releasing the message.

This works, I just had to make the member public rather than internal, so in my code I can assign the pointer returned from `ReceiveMessagesOnPollGroup` or `ReceiveMessagesOnConnection` in `IntPtr[] ppOutMessages` to this member and then calling `Release()` on the message works properly.

It might be worth to wrap this behavior around later to simplify the API for users, but it works without it!

I got the Steam Networking Sockets fully integrated with my project and so far it all works correctly, so it should be good! I haven't used all pieces of the API though, but the main ones seem good. Thanks again for working on this!
